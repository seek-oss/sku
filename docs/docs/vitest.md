# Vitest

[Vitest] is a testing framework that supports ESM out-of-the-box, integrates with the Vite ecosystem and has a similar API to Jest.
These features make it a great replacement for Jest in `sku` applications, especially given [Jest's current limitations with ESM].
Sku supports Vitest as an alternate to Jest since v15

**We recommend using vitest with the vite bundler and in an ESM project**, however these are not requirements.

## Migrating to Vitest

Migrating to Vitest should be fairly straightforward.
Before making any changes, it is highly recommend to read [the Vitest documentation](https://vitest.dev/guide/) to familiarise yourself with the API and features, as well as to better understand the differences between Jest and Vitest.

To enable Vitest in `sku`, first install the required dependencies:

```bash
pnpm add -D vitest @sku-lib/vitest
```

Then, configure [`testRunner`][test runner] in your `sku` config:

```typescript
// sku.config.ts
import type { SkuConfig } from 'sku';

export default {
  testRunner: 'vitest',
  ...
} satisfies SkuConfig;
```

`sku` will now invoke [the `vitest` CLI][Vitest CLI] instead of the `jest` CLI when running `sku test`.

#### Key differences between Vitest and Jest

Vitest has strong compatibility with the Jest API, however it still has differences that may affect your tests.
Differences that affect sku projects are listed here.
The full list is documented in the Vitest documentation: [Migrating from Jest to Vitest].

To automate most of the migration process, a codemod is available.
Note that additional changes may still be required after running this codemod.

```sh
pnpm dlx @sku-lib/codemod jest-to-vitest .
```

**Default Watch Mode**

`vitest` defaults to watch mode when running tests.
To run tests without watch mode you can use the `--run` flag:

```json
// package.json
{
  "scripts": {
    // ...
    "test": "sku test --run"
  }
}
```

Watch mode won't trigger in CI environments, so it's safe to omit the flag in your pipeline.

**Code Coverage Dependencies**

Vitest does not install code coverage dependencies by default.
To collect code coverage, pass the `--coverage` flag to the `sku test` command and install the `@vitest/coverage-v8` package:

```bash
pnpm add -D @vitest/coverage-v8
```

**Testing Library Matchers**

If your test setup file includes an import for `@testing-library/jest-dom`, you may need to change this to `@testing-library/jest-dom/vitest`:

```diff
// test-setup.ts
- import '@testing-library/jest-dom';
+ import '@testing-library/jest-dom/vitest';
```

**Globals Disabled**

Jest enables global APIs such as `it`, `describe`, `beforeAll`, etc., by default.
Vitest does not.
Sku adopts the Vitest defaults, meaning you will need to explicitly import these test functions.
This change results in a cleaner global namespace, better type safety, clearer dependencies and is more consistency with modern javascript practices.

```diff
+ import { describe, expect, it } from 'vitest';
```

Be aware that since globals are disabled, some common libraries like `testing-library` will not run auto DOM cleanup.
If using these libraries, you will need to add cleanup to your configured `setupTests` file.

```diff
import '@testing-library/jest-dom/vitest';

+ import { cleanup } from '@testing-library/react';
+ import { afterEach } from 'vitest';

+ afterEach(cleanup)
```

[Vitest]: https://vitest.dev/
[Jest's current limitations with ESM]: https://jestjs.io/docs/ecmascript-modules
[test runner]: ./docs/configuration.md#testrunner
[Vitest CLI]: https://vitest.dev/guide/cli.html
[codemod]: https://codemod.com/registry/jest-vitest
[Migrating from Jest to Vitest]: https://vitest.dev/guide/migration.html#jest
[Vitest globals]: https://vitest.dev/config/#globals
