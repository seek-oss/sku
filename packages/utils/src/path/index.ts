export { toPosixPath } from './toPosixPath.js';
