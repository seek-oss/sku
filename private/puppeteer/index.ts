export { getAppSnapshot } from './appSnapshot.ts';
export { appSnapshotSerializer } from './appSnapshotSerializer.ts';
export { htmlSnapshotSerializer } from './htmlSnapshotSerializer.ts';
export { cssSnapshotSerializer } from './cssSnapshotSerializer.ts';
