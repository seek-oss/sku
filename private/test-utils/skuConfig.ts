/**
 * A helper to make stable hashes for webpack.
 */
export const makeStableHashes = (config: Record<string, any>) => {
  // Addresses an issue where module IDs were different between local dev and CI
  config.optimization.chunkIds = 'natural';
  config.optimization.moduleIds = 'natural';

  // Disable hashes
  if (config.output.filename?.includes('[contenthash]')) {
    config.output.filename = '[name].js';
  }
  if (config.output.chunkFilename?.includes('[contenthash]')) {
    config.output.chunkFilename = '[name].chunk.js';
  }

  return config;
};

/**
 * A helper for vite to make stable hashes.
 */
export const makeStableViteHashes = () => ({
  build: {
    rollupOptions: {
      output: {
        chunkFileNames: '[name].js',
        assetFileNames: '[name][extname]',
        entryFileNames: '[name].js',
      },
    },
  },
});
