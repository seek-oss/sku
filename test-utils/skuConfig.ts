export const makeStableHashes = (config: Record<string, any>) => {
  // Addresses an issue where module IDs were different between local dev and CI
  config.optimization.chunkIds = 'natural';
  config.optimization.moduleIds = 'natural';

  // Disable hashes
  if (config.output.filename?.includes('[contenthash]')) {
    config.output.filename = '[name].js';
  }
  if (config.output.chunkFilename?.includes('[contenthash]')) {
    config.output.chunkFilename = '[name].chunk.js';
  }

  return config;
};
