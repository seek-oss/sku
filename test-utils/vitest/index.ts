export { getAppSnapshot } from './appSnapshot.ts';
