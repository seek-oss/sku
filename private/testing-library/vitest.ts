// Re-export vitest setup from cli-testing-library
import 'cli-testing-library/vitest';
