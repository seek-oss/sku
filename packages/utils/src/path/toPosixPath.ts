import path from 'node:path';

/**
 * Replaces all win32 path separators (\) with posix path separators (/)
 */
export const toPosixPath = (inputPath: string) =>
  inputPath.replaceAll(path.win32.sep, path.posix.sep);
