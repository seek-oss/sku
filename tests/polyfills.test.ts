import { describe, beforeAll, afterAll, it } from 'vitest';
import { getAppSnapshot } from '@sku-private/puppeteer';
import assert from 'node:assert/strict';
import path from 'node:path';
import {
  waitForUrls,
  runSkuScriptInDir,
  getPort,
  createCancelSignal,
} from '@sku-private/test-utils';
import skuConfig from '@sku-fixtures/polyfills/sku.config.ts';

import { createRequire } from 'node:module';

const require = createRequire(import.meta.url);

const appDir = path.dirname(
  require.resolve('@sku-fixtures/polyfills/sku.config.ts'),
);

assert(skuConfig.port, 'sku config has port');

describe('polyfills', () => {
  const args: Record<string, string[]> = {
    vite: ['--config', 'sku.config.vite.ts', '--experimental-bundler'],
  };

  describe.sequential.for(['webpack'])('bundler %s', async (bundler) => {
    const port = await getPort();
    const baseUrl = `http://localhost:${port}`;

    describe('build', () => {
      const { cancel, signal } = createCancelSignal();

      beforeAll(async () => {
        await runSkuScriptInDir('build', appDir, { args: args[bundler] });
        runSkuScriptInDir('serve', appDir, {
          args: ['--strict-port', `--port=${port}`],
          signal,
        });
        await waitForUrls(baseUrl);
      });

      afterAll(async () => {
        cancel();
      });

      it('should create valid app', async ({ expect }) => {
        const app = await getAppSnapshot({ url: baseUrl, expect });
        expect(app).toMatchSnapshot();
      });
    });

    describe('start', () => {
      const { cancel, signal } = createCancelSignal();
      const devServerUrl = `http://localhost:${skuConfig.port}`;

      beforeAll(async () => {
        runSkuScriptInDir('start', appDir, {
          signal,
        });
        await waitForUrls(devServerUrl);
      });

      afterAll(async () => {
        cancel();
      });

      it('should start a development server', async ({ expect }) => {
        const snapshot = await getAppSnapshot({ url: devServerUrl, expect });
        expect(snapshot).toMatchSnapshot();
      });
    });
  });
});
