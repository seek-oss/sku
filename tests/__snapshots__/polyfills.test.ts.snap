// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`polyfills > bundler webpack > build > should create valid app 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      polyfill
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/static/polyfills/runtime.js"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/static/polyfills/2.js"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/static/polyfills/main.js"
    >
  </head>
  <body>
    <div id="app">
      <div>
        App with a polyfill
      </div>
    </div>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
    >
      []
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
    >
      {"namedChunks":[]}
    </script>
    <script
      async
      data-chunk="main"
      src="/static/polyfills/runtime.js"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="/static/polyfills/2.js"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="/static/polyfills/main.js"
    >
    </script>
  </body>
</html>"
POST HYDRATE DIFFS: 
===================================================================
--- sourceHtml
+++ clientHtml
@@ -64,5 +64,11 @@
       src="/static/polyfills/main.js"
     >
     </script>
+    <p>
+      This node was injected by the 'injected' global function.
+    </p>
+    <p>
+      This node was injected by the '3rd-party-polyfill' dependency.
+    </p>
   </body>
 </html>
\\ No newline at end of file
`;

exports[`polyfills > bundler webpack > start > should start a development server 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      polyfill
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/runtime.js"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/2.js"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/main.js"
    >
  </head>
  <body>
    <div id="app">
      <div>
        App with a polyfill
      </div>
    </div>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
    >
      []
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
    >
      {"namedChunks":[]}
    </script>
    <script
      async
      data-chunk="main"
      src="/runtime.js"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="/2.js"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="/main.js"
    >
    </script>
  </body>
</html>"
POST HYDRATE DIFFS: 
===================================================================
--- sourceHtml
+++ clientHtml
@@ -64,5 +64,11 @@
       src="/main.js"
     >
     </script>
+    <p>
+      This node was injected by the 'injected' global function.
+    </p>
+    <p>
+      This node was injected by the '3rd-party-polyfill' dependency.
+    </p>
   </body>
 </html>
\\ No newline at end of file
`;
